### PR TITLE
fix(RHTAPBUGS-959): run internal requests in parallel in rh-sign-image

### PR DESCRIPTION
**Pull request by mmalina**
Up to 4 IRs will run in parallel. This will improve the performance of the task.

Note that this depends on https://github.com/redhat-appstudio/release-service-utils/pull/98

### CODE REVIEW

File: tasks/rh-sign-image/rh-sign-image.yaml
Comment (mmalina) [inline tasks/rh-sign-image/rh-sign-image.yaml]: Note: This needs to be changed before merging.

File: tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml
Comment (johnbieren) [inline tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: I'm probably missing something obvious, but what is making this test case time out?

File: tasks/rh-sign-image/tests/mocks.sh
Comment (johnbieren) [inline tasks/rh-sign-image/tests/mocks.sh:11]: do we want to extract this to something in the .github directory for other IR tasks to use? It seems pretty generic. It could be done as a followup ticket+pr

File: tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml
Comment (mmalina) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: It's because the `requestTimeout` parameter of the task is set to `1`. And this is passed to the `wait-for-ir` call. So the mock would mark the IR as passed in 10 seconds, but it will time out before that happens.

File: tasks/rh-sign-image/tests/mocks.sh
Comment (mmalina) [reply tasks/rh-sign-image/tests/mocks.sh:11]: Hmm, I'm not opposed to it, but I'm not sure where it would fit. Any suggestions? And what makes this different than other mocks we have? The IR name extraction maybe? Anyway, I would probably treat this the same way as the rest - meaning that this will be copied to other places as needed.

Maybe the `set_ir_status` function would be nice to have ready. But with how this works, it would probably have to be baked in the utils image. Which would make it opaque, so not sure if it's a good idea (and also, it's for testing, so it doesn't seem great to have it in the image).

File: tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml
Comment (johnbieren) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: I see. It makes sense now. It may be worth writing pretty much exactly that in the description, but not required. I'll leave it up to you

File: tasks/rh-sign-image/tests/mocks.sh
Comment (johnbieren) [reply tasks/rh-sign-image/tests/mocks.sh:11]: If we go with a similar change for the other IR tasks, I'd think we need something just like this in them all. Duplicating it is fine, I was just trying to think ahead. We can talk about it as we copy this logic in other task tests

**Review by johnbieren — DISMISSED** [review]

File: tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml
Comment (mmalina) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: ok, added.

File: tasks/rh-sign-image/tests/mocks.sh
Comment (mmalina) [reply tasks/rh-sign-image/tests/mocks.sh:11]: Yeah, I agree. Maybe we can think of a way to deduplicate when we add this to other tasks. Or we just copy it.

File: tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml
Comment (johnbieren) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: did you remember to push? Doesn't seem like there are any new updates

Comment (mmalina) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: oops, thanks. turns out I accidentally pushed it to the main repo and not my fork. updated now.

Comment (johnbieren) [reply tasks/rh-sign-image/tests/test-rh-sign-image-timeout.yaml]: looks good, thanks

**Review by johnbieren — DISMISSED** [review]
LGTM but still waiting on the release-service-utils PR before merging

File: tasks/rh-sign-image/rh-sign-image.yaml
Comment (theflockers) [inline tasks/rh-sign-image/rh-sign-image.yaml:99]: I wonder how it would behave in case for instance: You have 5 requests, so 4 would be running in parallel. The 5th would start only when all 4 finishes, right? What if 3 finishes and one gets stuck? It would wait until all requests are done to start the 5th, correct?

Comment (mmalina) [reply tasks/rh-sign-image/rh-sign-image.yaml:99]: > You have 5 requests, so 4 would be running in parallel. The 5th would start only when all 4 finishes, right?

Correct.

> What if 3 finishes and one gets stuck? It would wait until all requests are done to start the 5th, correct?

If it gets stuck, eventually the `wait-for-ir` script will time out and the pipeline fails. The fifth ir will never get created.

> It would wait until all requests are done to start the 5th, correct?

Yes, it won't proceed to the next irs until the previous batch is done.

Comment (theflockers) [reply tasks/rh-sign-image/rh-sign-image.yaml:99]: I see. I think we should move forward this way but IMHO this should be more like a queue format - so we would always have "slots" available to create the next request when  another one is finished. I will write a proposal of this in the course of this week (using spare time) if you and @davidmogar  don't mind.

**Review by theflockers — APPROVED** [review]

File: tasks/rh-sign-image/rh-sign-image.yaml
Comment (mmalina) [reply tasks/rh-sign-image/rh-sign-image.yaml:99]: Yeah, I thought about that. But it would make it more complex and I'm not sure if there is much value in doing that. At the end of the day the only reason to reduce the throughput a bit is so that we don't overwhelm the services we use, like IIB or RADAS. And the current design allows us to regulate this using the concurrencyLimit parameter. For example, using your suggested approach with concurrencyLimit=4 could have the same performance (and load on the services) as concurrencyLimix=5 with my approach. And also, the IRs in a given tasks will typically take a similar time to complete - there shouldn't be a big spread.

**Review by johnbieren — APPROVED** [review]

**Pull request merged**
